src/execution.ts

```diff
@@ -604,23 +610,41 @@ function compileLeafType(
  * Compile a node of object type.
  * @param {CompilationContext} context
  * @param {GraphQLObjectType} type type of the node
+ * @param fieldNodes fieldNodes array with the nodes references
  * @param originPaths originPaths path in the parent object from where to fetch results
  * @param destinationPaths path in the where to write the result
  * @param responsePath response path until this node
- * @param fieldMap fieldNodes array with the nodes references
+ * @param errorDestination Path for error array
+ * @param fieldMap map of fields to fieldNodes array with the nodes references
  * @param alwaysDefer used to force the field to be resolved with a resolver ala graphql-js
  * @returns {string}
  */
 function compileObjectType(
   context: CompilationContext,
   type: GraphQLObjectType,
+  fieldNodes: FieldNode[],
   originPaths: string[],
   destinationPaths: string[],
   responsePath: ObjectPath | undefined,
+  errorDestination: string,
   fieldMap: { [key: string]: FieldNode[] },
   alwaysDefer: boolean
 ): string {
-  let body = `{`;
+  let body = "(";
+  if (typeof type.isTypeOf === "function" && !alwaysDefer) {
+    context.dependencies.set(type.name + "IsTypeOf", type.isTypeOf);
+    body += `!${type.name}IsTypeOf(${originPaths.join(
+      "."
+    )}) ? (${errorDestination}.push(${createErrorObject(
```

Comment:
this could be an if..else. what do you think ... both branches are used as statements rather than expressions.

Reply:
This is inside a giant expression so I can't really change it in this PR.